### ISSUE TITLE
Not working when cross compiling

### ISSUE THREAD

**Issue by semigodking** [alehaa/CMake-argp#1]
Is it possible to enhance this script to support cross compiling?

**Comment by alehaa** [comment]
Can you provide some details about this issue?

The modules uses `find_library` and `check_function_exists`, which use the compiler ans root path of the toolchain file.

**Comment by semigodking** [comment]
Hi Alex, I attached an sample project using this module.  I was trying to compile it for WIN32 by MINGW32.  cmake can not find the argp lib. Manual linking with argp success. I am thinking if this issue is caused by checking function argp_parse fails.

[sample.zip](https://github.com/alehaa/CMake-argp/files/812635/sample.zip)

Here is screen capture.

```
[kitty@default build]$ mingw32-cmake ../
-- Looking for argp_parse
-- Looking for argp_parse - not found
-- Could NOT find argp (missing:  ARGP_LIBRARIES)
-- Configuring done
-- Generating done
-- Build files have been written to: /home/kitty/sample/build
```
```
[kitty@default build]$ mingw32-make
Scanning dependencies of target test
[100%] Building C object CMakeFiles/test.dir/main.c.obj
Linking C executable test.exe
CMakeFiles/test.dir/objects.a(main.c.obj):main.c:(.text.startup+0x3c): undefined reference to `argp_parse'
collect2: error: ld returned 1 exit status
make[2]: *** [test.exe] Error 1
make[1]: *** [CMakeFiles/test.dir/all] Error 2
make: *** [all] Error 2
```
```
[kitty@default build]$ mingw32-make VERBOSE=1
/usr/bin/cmake -H/home/kitty/sample -B/home/kitty/sample/build --check-build-system CMakeFiles/Makefile.cmake 0
/usr/bin/cmake -E cmake_progress_start /home/kitty/sample/build/CMakeFiles /home/kitty/sample/build/CMakeFiles/progress.marks
make -f CMakeFiles/Makefile2 all
make[1]: Entering directory `/home/kitty/sample/build'
make -f CMakeFiles/test.dir/build.make CMakeFiles/test.dir/depend
make[2]: Entering directory `/home/kitty/sample/build'
cd /home/kitty/sample/build && /usr/bin/cmake -E cmake_depends "Unix Makefiles" /home/kitty/sample /home/kitty/sample /home/kitty/sample/build /home/kitty/sample/build /home/kitty/sample/build/CMakeFiles/test.dir/DependInfo.cmake --color=
make[2]: Leaving directory `/home/kitty/sample/build'
make -f CMakeFiles/test.dir/build.make CMakeFiles/test.dir/build
make[2]: Entering directory `/home/kitty/sample/build'
Linking C executable test.exe
/usr/bin/cmake -E cmake_link_script CMakeFiles/test.dir/link.txt --verbose=1
/usr/bin/cmake -E remove -f CMakeFiles/test.dir/objects.a
/usr/bin/i686-w64-mingw32-ar cr CMakeFiles/test.dir/objects.a @CMakeFiles/test.dir/objects1.rsp
/usr/bin/i686-w64-mingw32-gcc  -O3 -DNDEBUG    -Wl,--whole-archive CMakeFiles/test.dir/objects.a -Wl,--no-whole-archive  -o test.exe -Wl,--out-implib,libtest.dll.a -Wl,--major-image-version,0,--minor-image-version,0  -L/home/kitty/argp-standalone  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32
CMakeFiles/test.dir/objects.a(main.c.obj):main.c:(.text.startup+0x3c): undefined reference to `argp_parse'
collect2: error: ld returned 1 exit status
make[2]: *** [test.exe] Error 1
make[2]: Leaving directory `/home/kitty/sample/build'
make[1]: *** [CMakeFiles/test.dir/all] Error 2
make[1]: Leaving directory `/home/kitty/sample/build'
make: *** [all] Error 2
```
```
[kitty@default build]$ /usr/bin/i686-w64-mingw32-gcc  -O3 -DNDEBUG    -Wl,--whole-archive CMakeFiles/test.dir/objects.a -Wl,--no-whole-archive  -o test.exe -Wl,--out-implib,libtest.dll.a -Wl,--major-image-version,0,--minor-image-version,0  -L/home/kitty/argp-standalone  -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -largp
```
```
[kitty@default build]$ ls -al test.exe
-rwxrwxr-x 1 kitty kitty 382586 Mar  2 10:48 test.exe
[kitty@default build]$
```

**Comment by alehaa** [comment]
I think I found the bug and it seems not to be cross-compile specific.

In the attached zip file I found the `include_directories()` command for  argp-standalone. As my cmake module doesn't search for the header it won't be found in non-standard directories, so it will crash on your system.

I'll try to enhance the script to search for the header file first.

**Comment by semigodking** [comment]
Good to know. Thank you!

**Comment by alehaa** [comment]
@semigodking After I looked into your code once again, could you please do the following patch on `Findargp.cmake` and give me the output after running? Please delete the caches before (i.e. delete the build directory and run cmake again).

```diff
--- a/cmake/Findargp.cmake
+++ b/cmake/Findargp.cmake
@@ -25,10 +25,11 @@ include(CheckFunctionExists)
 
 # Check if argp is included in libc:
 set(CMAKE_REQUIRED_QUIET_SAVE ${CMAKE_REQUIRED_QUIET})
-set(CMAKE_REQUIRED_QUIET TRUE)
+# set(CMAKE_REQUIRED_QUIET TRUE)
 
 check_function_exists("argp_parse" ARGP_IN_LIBC)
 if (ARGP_IN_LIBC)
+       message(STATUS "argp: found in libc")
        set(ARGP_LIBRARIES "c" CACHE STRING "ARGP libraries.")
 
 elseif (NOT ARGP_IN_LIBC)
@@ -36,6 +37,7 @@ elseif (NOT ARGP_IN_LIBC)
 
        find_library(ARGP_LIB "argp")
        mark_as_advanced(ARGP_LIB)
+       message(STATUS "argp: lib: ${ARGP_LIB}")
        if (ARGP_LIB)
                set(CMAKE_REQUIRED_LIBRARIES "${ARGP_LIB}")
                check_function_exists("argp_parse" ARGP_EXTERNAL)
```

**Comment by semigodking** [comment]
@alehaa Hi Alex, here is cmake output. 

```
[kitty@default build]$ rm -rf *
[kitty@default build]$ mingw32-cmake -d ../
-- The C compiler identification is GNU 4.9.2
-- Check for working C compiler: /usr/bin/i686-w64-mingw32-gcc
-- Check for working C compiler: /usr/bin/i686-w64-mingw32-gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Looking for argp_parse
-- Looking for argp_parse - not found
-- argp: lib: ARGP_LIB-NOTFOUND
-- Could NOT find argp (missing:  ARGP_LIBRARIES)
-- Configuring done
-- Generating done
CMake Warning:
  Manually-specified variables were not used by the project:

    BUILD_SHARED_LIBS
    CMAKE_INSTALL_LIBDIR
    INCLUDE_INSTALL_DIR
    SHARE_INSTALL_PREFIX
    SYSCONF_INSTALL_DIR


-- Build files have been written to: /home/kitty/sample/build
```

**Comment by alehaa** [comment]
I thought something like this would happen. ;) It doesn't find your argp library. Try adding `-D CMAKE_LIBRARY_PATH=<path to argp>` and it should find argp.

**Comment by semigodking** [comment]
No, it still can not find argp.

**Comment by alehaa** [comment]
Ok. Your CMake does not find the library. Except for the header to be found, this is an issue with `find_library`. Did you try the other variables [in the manual](https://cmake.org/cmake/help/v3.0/command/find_library.html)? Maybe the variables are different for `mingw32-cmake`?

**Comment by alehaa** [comment]
@semigodking Please check the version in rework branch. Again, please add some debugging and remove the `MINGW` argp settings in your root CMake file: 

```diff
--- a/cmake/Findargp.cmake
+++ b/cmake/Findargp.cmake
@@ -25,13 +25,14 @@ include(CheckFunctionExists)
 
 # Do the following checks for header, library and argp functions quietly. Only
 # print the result at the end of this file.
-set(CMAKE_REQUIRED_QUIET TRUE)
+# set(CMAKE_REQUIRED_QUIET TRUE)
 
 
 # First search the argp header file. If it is not found, any further steps will
 # fail.
 find_path(ARGP_INCLUDE_PATH "argp.h")
 if (ARGP_INCLUDE_PATH)
+       message("argp include: ${ARGP_INCLUDE_PATH}")
        set(CMAKE_REQUIRED_INCLUDES "${ARGP_INCLUDE_PATH}")
 
        # Find the required argp library. argp may be shipped together with libc (as
@@ -43,6 +44,7 @@ if (ARGP_INCLUDE_PATH)
                # otherwise libc doesn't ship it.
                check_function_exists("argp_parse" ARGP_IN_LIBC)
                if (ARGP_IN_LIBC)
+                       message("argp in libc")
                        set(ARGP_LIBRARIES "c" CACHE STRING
                            "Libraries required for using argp.")
 
@@ -51,6 +53,7 @@ if (ARGP_INCLUDE_PATH)
                else ()
                        find_library(ARGP_LIBRARIES "argp")
                        if (ARGP_LIBRARIES)
+                               message("argp: lib: ${ARGP_LIBRARIES}")
                                set(CMAKE_REQUIRED_LIBRARIES "${ARGP_LIBRARIES}")
                                check_function_exists("argp_parse" ARGP_EXTERNAL)
                                if (NOT ARGP_EXTERNAL)
```

**Comment by semigodking** [comment]
Hi Alex,
Thank you for your help. Even with the new patch, argp could not be found.
I tried many ways. I found that the include dirs set by include_directories
make no sense at all in mingw32-cmake. I give up and do not want to waste
your time. Thanks again.

Regards,
Zhuofei Wang

**Comment by semigodking** [comment]
Hi Alex,
I finally made it working by appending argument CMAKE_FIND_ROOT_PATH_BOTH to find_path and find_library. And, here is how to invoke find_package(argp). One thing I do not understand is that why I need to append the same path to CMAKE_INCLUDE_PATH  as well as include_directories. It looks like that CMAKE_INCLUDE_PATH is used for finding path while include_directories is used for adding additional include path to compiler.

```cmake
if (${MINGW})
   list(APPEND CMAKE_INCLUDE_PATH "~/argp-standalone")
   list(APPEND CMAKE_LIBRARY_PATH "~/argp-standalone")
   include_directories("~/argp-standalone")
   link_directories ("~/argp-standalone")
endif ()

find_package(argp)
```
Findargp.cmake
```cmake
# This file is part of CMake-argp.
#
# CMake-argp is free software: you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License as published by the Free
# Software Foundation, either version 3 of the License, or (at your option) any
# later version.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE. See the GNU General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License along
# with this program. If not, see
#
#  http://www.gnu.org/licenses/
#
#
# Copyright (c)
#   2016 Alexander Haase <ahaase@alexhaase.de>
#

include(FindPackageHandleStandardArgs)
include(CheckFunctionExists)


# Check if argp is included in libc:
set(CMAKE_REQUIRED_QUIET_SAVE ${CMAKE_REQUIRED_QUIET})
# set(CMAKE_REQUIRED_QUIET TRUE)

# First search the argp header file. If it is not found, any further steps will
# fail.
FIND_PATH(ARGP_INCLUDE_PATH argp.h CMAKE_FIND_ROOT_PATH_BOTH )
message("argp include: ${ARGP_INCLUDE_PATH}")
if (ARGP_INCLUDE_PATH)
        set(CMAKE_REQUIRED_INCLUDES "${ARGP_INCLUDE_PATH}")

        check_function_exists("argp_parse" ARGP_IN_LIBC)
        if (ARGP_IN_LIBC)
                message(STATUS "argp in libc")
                set(ARGP_LIBRARIES "c" CACHE STRING "ARGP libraries.")

        elseif (NOT ARGP_IN_LIBC)
                unset(ARGP_IN_LIBC CACHE)

                find_library(ARGP_LIB "argp" CMAKE_FIND_ROOT_PATH_BOTH)
                mark_as_advanced(ARGP_LIB)
                message(STATUS "argp: lib: ${ARGP_LIB}")
                if (ARGP_LIB)
                        set(CMAKE_REQUIRED_LIBRARIES "${ARGP_LIB}")
                        check_function_exists("argp_parse" ARGP_EXTERNAL)
                        if (ARGP_EXTERNAL)
                                set(ARGP_LIBRARIES "${ARGP_LIB}" CACHE STRING "ARGP libraries.")
                        endif ()
                endif ()
        endif ()
endif ()

set(CMAKE_REQUIRED_QUIET ${CMAKE_REQUIRED_QUIET_SAVE})


find_package_handle_standard_args(argp
        FOUND_VAR ARGP_FOUND
        REQUIRED_VARS ARGP_LIBRARIES
)
```

**Comment by semigodking** [comment]
another solution:
Keep your patched Findargp.cmake unchanged.
Invoke find_package(argp) like below:
```cmake
if (${MINGW})
   set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY BOTH)
   set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE BOTH)
endif ()

find_package(argp REQUIRED)
include_directories(${ARGP_INCLUDE_PATH})
set (EXTRA_LIBS ${EXTRA_LIBS} ${ARGP_LIBRARIES})
```
And invoke mingw32-cmake:
```shell
mingw32-cmake -DCMAKE_LIBRARY_PATH=~/argp-standalone -DCMAKE_INCLUDE_PATH=~/argp-standalone ../
```

**Comment by alehaa** [comment]
I think I've figgured out why this doesn't work.

I think your argp header and library are stored in `~/argp-standalone`, but not in independent subdirectories named `lib` and `include`, right? If the files are stored in these subdirectories, it should be sufficient to call `minge32-cmake` with `-DCMAKE_PREFIX_PATH=~/argp-standalone`.

But I still don't know why you have to set `CMAKE_FIND_ROOT_PATH_MODE_*`to `BOTH`, as the [manual](https://cmake.org/cmake/help/v3.0/command/find_library.html) says, it'll search in `CMAKE_LIBRARY_PATH`, too.

**Comment by semigodking** [comment]
According to cmake manual, without setting CMAKE_FIND_ROOT_PATH_MODE_* to
BOTH,cmake searches target sysroot for files. When they are set to BOTH,
cmake searches both target sysroot and host for files. So, unless I have
argp-standalone compiled and installed into mingw' sysroot, I have to
instruct cmake to search host directories.

**Comment by alehaa** [comment]
Ok. Am I right that this an issue of your installation but not of my script?`So can I close this issue now and merge the patches for finding the header?

**Comment by semigodking** [comment]
Yes, you are right. I would like to suggest you keep find_path for header part. this part makes finding and including argp header easier.

**Comment by alehaa** [comment]
That was the plan ;)

**Comment by alehaa** [comment]
Merged the fixes in 5523dc6b7de7810389162ee7c55b8e1c4b3f2831.
